profiles.wordpress.org/public_html/wp-content/plugins/wporg-5ftf-activity-handler/wporg-5ftf-activity-handler.php

```diff
@@ -0,0 +1,65 @@
+<?php
+/*
+Plugin Name: WordPress.org 5ftf Activity Handler
+Plugin URI: http://wordpress.org
+License: GPL2
+Version: 1.1
+Description: Handles saving of last 5ftf contribution.
+*/
+
+defined( 'ABSPATH' ) or die();
+
+if ( ! class_exists( 'WPOrg_5ftf_Activity_Handler' ) ) {
+
+	class WPOrg_5ftf_Activity_Handler {
+
+		/**
+		 *  Name of meta key that tracks last contribution as a unix timestamp.
+		 *
+		 * @var string
+		 */
+		const last_contribution_meta_key = 'wporg_5ftf_last_contribution';
+
+		/**
+		 * Constructor.
+		 */
+		public function __construct() {
+			add_filter( 'bp_activity_add', array( $this, 'handle_contribution' ) );
+		}
+
+		/**
+		 * Saves contribution if it qualifies as a contribution.
+		 */
+		public function handle_contribution( $args ) {
+			if ( self::is_5ftf_contribution( $args['type'] ) ) {
+				self::update_last_contribution_meta( $args['user_id'] );
+			}
+		}
+
+		/**
+		 * Returns whether action is considered a contribution.
+		 */
+		public function is_5ftf_contribution( $action ) {
+			$wordpress_actions = array( 'blog_post_create' );
+			$wordcamp_actions  = array( 'wordcamp_speaker_add', 'wordcamp_organizer_add' );
+
+			$valid_actions = array_merge( $wordpress_actions, $wordcamp_actions );
+
```

Comment:
There could be more stuff here for the complex cases, like 

```php
if ( 'forum_topic_reply' === $action ) {
    return is_5ftf_forum_contribution( $action, $forum_id, $etc );
}
```

... which calls something like

```php
function is_5ftf_forum_contribution( $action, $forum, $etc ) {
   // if creating a topic, return false
   // if replying to their own topic, return false
   // if plugin/theme forum, return false
   // return true
}
```

there might be a better way, that's just what feels the most straight-forward at the moment. what do you think?

Reply:
I'm a bit concerned that it may become difficult to read if we are doing a lot of source information validation (like checking for a specific forum `id`, running extra queries and running additional logical statements). 

My proposed alternative, which I was trying to portray in [our conversation on the other pull request](https://github.com/WordPress/wordpress.org/pull/75#issuecomment-1168090751), is that within the main profile activity handler, within the specific handle functions (ie: `handle_forum_activity`), we could create new `type`s that we can filter for in this plugin. This doesn't remove the need to check for source information, that would still have to happen in the main profile activity handler, but it at least keeps source information and validation where we already do some of that and have access to _all_ the activity information, not only the information that is passed to `bp_activity_add`. 

Additionally, this plugin will probably become the go-to reference to determine what is considered a contribution and what is not as it will be the only place where the safelist will exist, so it's important we can read this file quickly & easily. 

So, as an example:

**Case**:
"For instance, posting replies to someone else's post in the Fixing WP forums is a 5ftf contribution, but posting in a plugin support forum isn't."

**Example**:
[Source](https://github.com/WordPress/wordpress.org/blob/2d3cae5f929c3c20acbeb2e6c6dd74ea4f09dc02/profiles.wordpress.org/public_html/wp-content/plugins/wporg-profiles-activity-handler/wporg-profiles-activity-handler.php#L306)
```
private function handle_forum_activity() {
  $user = self::get_user( $_POST['user'] );
  $type = '';

  // replying to someone else's post in fixing wp forum
  if() {
   $type = 'forum_fixing_WP_reply`
  }

  ...
  $args = array(
    'component'         => 'forums',
    'type'              => $type,
  );

  ...
}
```

Then in this plugin, look for that type.
```
$forum_actions  = array( 'forum_fixing_WP_reply' );
```

I'm sure there's a better convention for the `type` string, but it's just an example. 

Now, I'm not sure this is the best, and I don't know what implications it would have to change the type... Maybe there is downstream logic that would break if we did it. I would have to investigate more.

Reply:
In looking at the GitHub notification code, I think we can apply a filter and handle it here. We'll need logical statements for that in this plugin anyways, so yeah. Your suggestion seems to be the best approach. 👍

Reply:
> GitHub notification code... apply a filter and handle it here

🤔 The GitHub code runs on w.org, but the activity handler runs on buddypress/profiles; are you thinking this plugin would run on both of those networks?

I like that idea, since the same function could be used for DRYness.

Reply:
> it may become difficult to read if we are doing a lot of source information validation ...  

Yeah, that's a good concern. I'm not sure about changing the `type` that gets passed to `bp_activity_add()`, though, since we'd lose the ability to perform actions on them all as a group (eg, `wp bp activity list --type=topic_forum_reply`). We'd also have to retroactively update all the existing items with their new types. It seems like it might be using `type` for something it wasn't intended for.

We could add that extra data as a meta item on the activity though, and then `if_5ftf_contribution()` could lookup that value for its logic.

We could maybe use activity metadata instead

Reply:
Er, extra meta fields might not be great either, since it'd add a lot of data to the db, and it feels a bit unnecessary.

Functions like `handle_forum_activity()` are already pretty long and do too many things, so adding more to them might not help avoid the "difficult to read" problem.

🤔 

I like your point about keeping the 5ftf check in the same place as the other validation, where we have access to all of the data. At the same time it feels nice to have that stuff in a separate plugin, so that each is only doing "one thing".

Is there anything we'd need for the `is_5ftf_contribution()` calculation that doesn't exist in the database? It doesn't seem like a big deal for query for a few things if they're already saved as activity meta data. There might be cases where we'll have to do that anyway, like getting the author of the forum post that is being replied to (to check if it's the same as the person making the reply).

I'm still open to exploring stuff like #75 or adding the logic to the `handle_...` functions too, so feel free to push back if you disagree.

Reply:
I will **not** be exploring the `type` approach. I agree with your assessment.

Reply:
> 🤔 The GitHub code runs on w.org, but the activity handler runs on buddypress/profiles; are you thinking this plugin >would run on both of those networks?
>
> I like that idea, since the same function could be used for DRYness.

Yeah, I'm thinking we could also handle trac in the same way. Only this file will, provided I don't run into any blockers, save to the meta field.